### PR TITLE
Merge redacted send/recv with master

### PR DESCRIPTION
**Pull request by pcd1193182**

### CODE REVIEW

File: cmd/zfs/zfs_main.c [hunk 1]
```diff
@@ -4127,7 +4127,7 @@ zfs_do_send(int argc, char **argv)
 
 	struct option long_options[] = {
 		{"replicate",	no_argument,		NULL, 'R'},
-		{"redact-bookmark",	required_argument, NULL, REDACT_OPT},
+		{"redact",	required_argument,	NULL, 'd'},
```
Comment (ahrens) [inline cmd/zfs/zfs_main.c]: REDACT_OPT isn't used any longer.  It looks like it also isn't present upstream, so we should remove it.

File: include/libzfs_core.h [hunk 2]
```diff
@@ -68,6 +68,7 @@ int lzc_initialize(const char *, pool_initialize_func_t, nvlist_t *,
     nvlist_t **);
 int lzc_trim(const char *, pool_trim_func_t, uint64_t, boolean_t,
     nvlist_t *, nvlist_t **);
+int lzc_redact(const char *, const char *, nvlist_t *);
```
Comment (ahrens) [inline include/libzfs_core.h]: remove the old declaration above

File: man/man5/zpool-features.5 [hunk 3]
```diff
@@ -645,6 +666,46 @@ The upgrade process runs in the background and may take a while to complete
 for the filesystems containing a large number of files.
 .RE
 
+.sp
+.ne 2
+.na
+\fB\fBredaction_bookmarks\fR\fR
```
Comment (ahrens) [inline man/man5/zpool-features.5]: remove this section from its old location

File: man/man5/zpool-features.5 [hunk 4]
```diff
@@ -645,6 +666,46 @@ The upgrade process runs in the background and may take a while to complete
...
+creates redaction bookmarks, which store the list of blocks redacted by the
+send that created them.  For more information about redacted send,
+see \fBzfs\fR(8).
+
+.RE
+
+.sp
+.ne 2
+.na
+\fB\fBredacted_datasets\fR\fR
```
Comment (ahrens) [inline man/man5/zpool-features.5]: remove this section from its old location

**Comment by pcd1193182** [comment]
@ahrens I have those running now, I'll update when they've finished.

**Comment by ahrens** [comment]
This PR should make our send/recv code identical to upstream, I think.  I ran `git diff origin/54/head..30af21b02569ac192f52ce6e6511015f8a8d5729` and looked for send/recv related differences.  Could you take a look at these diffs and see if we should apply them as part of this PR?
```
diff --git a/include/sys/dmu_recv.h b/include/sys/dmu_recv.h
index 3f6e6ea9de..1a7347d66e 100644
--- a/include/sys/dmu_recv.h
+++ b/include/sys/dmu_recv.h
@@ -31,6 +31,7 @@

 #include <sys/inttypes.h>
 #include <sys/dsl_bookmark.h>
+#include <sys/dsl_dataset.h>
 #include <sys/spa.h>
 #include <sys/objlist.h>
 #include <sys/dsl_bookmark.h>
@@ -45,12 +46,11 @@ typedef struct dmu_recv_cookie {
        const char *drc_tosnap;
        boolean_t drc_newfs;
        boolean_t drc_byteswap;
-       boolean_t drc_raw;
        uint64_t drc_featureflags;
        boolean_t drc_force;
        boolean_t drc_resumable;
+       boolean_t drc_raw;
        boolean_t drc_clone;
-       boolean_t drc_illumos; // keep this
        boolean_t drc_spill;
        struct avl_tree *drc_guid_to_ds_map;
        nvlist_t *drc_keynvl;
diff --git a/include/sys/dmu_send.h b/include/sys/dmu_send.h
index 9cbb7afa9e..2f3dfc39fd 100644
--- a/include/sys/dmu_send.h
+++ b/include/sys/dmu_send.h
@@ -48,7 +48,6 @@ struct drr_begin;
 struct avl_tree;
 struct dmu_replay_record;
 struct dmu_send_outparams;
-
 int
 dmu_send(const char *tosnap, const char *fromsnap, boolean_t embedok,
     boolean_t large_block_ok, boolean_t compressok, boolean_t rawok,
diff --git a/include/sys/dsl_bookmark.h b/include/sys/dsl_bookmark.h
index 915887ee46..05ebb76d23 100644
--- a/include/sys/dsl_bookmark.h
+++ b/include/sys/dsl_bookmark.h
@@ -35,6 +35,8 @@ typedef struct zfs_bookmark_phys {
        uint64_t zbm_guid;              /* guid of bookmarked dataset */
        uint64_t zbm_creation_txg;      /* birth transaction group */
        uint64_t zbm_creation_time;     /* bookmark creation time */
+
+       /* the following fields are reserved for redacted send / recv */
        uint64_t zbm_redaction_obj;     /* redaction list object */
        uint64_t zbm_flags;             /* ZBM_FLAG_* */
        uint64_t zbm_referenced_bytes_refd;
diff --git a/man/man8/zfs.8 b/man/man8/zfs.8
index deaac16f73..0a23466b7f 100644
--- a/man/man8/zfs.8
+++ b/man/man8/zfs.8
@@ -212,18 +212,12 @@
 .Ar snapshot
 .Nm
 .Cm send
-.Fl -redact Ar redaction_bookmark
-.Op Fl DLPcenpv
-.Op Fl i Ar snapshot Ns | Ns Ar bookmark
-.Ar snapshot
-.Nm
-.Cm send
 .Op Fl Penv
 .Fl t Ar receive_resume_token
 .Nm
 .Cm receive
 .Op Fl Fhnsuv
-.Op Fl o Ar origin=<snapshot>
+.Op Fl o Sy origin Ns = Ns Ar snapshot
 .Op Fl o Ar property Ns = Ns Ar value
 .Op Fl x Ar property
 .Ar filesystem Ns | Ns Ar volume Ns | Ns Ar snapshot
diff --git a/module/zfs/dnode_sync.c b/module/zfs/dnode_sync.c
index 9a5479fe44..d3acf1baae 100644
--- a/module/zfs/dnode_sync.c
+++ b/module/zfs/dnode_sync.c
@@ -32,8 +32,6 @@
 #include <sys/dmu_tx.h>
 #include <sys/dmu_objset.h>
 #include <sys/dmu_recv.h>
-#include <sys/dmu_send.h>
-#include <sys/dmu_recv.h>
 #include <sys/dsl_dataset.h>
 #include <sys/spa.h>
 #include <sys/range_tree.h>
diff --git a/tests/zfs-tests/tests/functional/cli_root/zfs_send/zfs_send-b.ksh b/tests/zfs-tests/tests/functional/cli_root/zfs_send/zfs_send-b.ksh
index 3fdd9b63c6..ab2a9508e4 100755
--- a/tests/zfs-tests/tests/functional/cli_root/zfs_send/zfs_send-b.ksh
+++ b/tests/zfs-tests/tests/functional/cli_root/zfs_send/zfs_send-b.ksh
@@ -62,6 +62,10 @@ for opt in ${opts[@]}; do
        log_must eval "zfs send -b$opt -i $SENDFS@s1 $SENDFS@s2 > /dev/null"
        log_must eval "zfs send -b$opt -I $SENDFS@s1 $SENDFS@s2 > /dev/null"
 done
+for opt in ${opts[@]}; do
+       log_mustnot eval "zfs send -b$opt $SENDFS > /dev/null"
+       log_mustnot eval "zfs send -b$opt $SENDFS#bm > /dev/null"
+done

 # Do 3..6 in a loop to verify various combination of "zfs send" options
 typeset opts=("" "p" "R" "pR" "cew")
diff --git a/tests/zfs-tests/tests/functional/redacted_send/redacted.kshlib b/tests/zfs-tests/tests/functional/redacted_send/redacted.kshlib
index c9c9063963..d3d219bb9b 100644
--- a/tests/zfs-tests/tests/functional/redacted_send/redacted.kshlib
+++ b/tests/zfs-tests/tests/functional/redacted_send/redacted.kshlib
@@ -245,7 +245,7 @@ function redacted_cleanup
                datasetexists $ds && log_must zfs destroy -R $ds
        done

-       echo "0" > /sys/module/zfs/parameters/zfs_allow_redacted_dataset_mount
+       log_must set_tunable32 zfs_allow_redacted_dataset_mount 0
        rm -f $(get_prop mountpoint $POOL)/tmp/*
 }

diff --git a/tests/zfs-tests/tests/functional/redacted_send/redacted_mounts.ksh b/tests/zfs-tests/tests/functional/redacted_send/redacted_mounts.ksh
index b520751336..b24f5ccad2 100755
--- a/tests/zfs-tests/tests/functional/redacted_send/redacted_mounts.ksh
+++ b/tests/zfs-tests/tests/functional/redacted_send/redacted_mounts.ksh
@@ -77,7 +77,7 @@ log_must zfs redact $sendvol@snap book2 $clonevol@snap
 log_must eval "zfs send --redact book2 $sendvol@snap >$stream"
 log_must eval "zfs receive $recvvol <$stream"
 [[ -b $recv_vol_file ]] && log_fail "Volume device file should not exist."
-echo "1" > /sys/module/zfs/parameters/zfs_allow_redacted_dataset_mount
+log_must set_tunable32 zfs_allow_redacted_dataset_mount 1
 log_must zpool export $POOL2
 log_must zpool import $POOL2
 udevadm settle
```

**Comment by pcd1193182** [comment]
We've got passing zfs-test, blackbox, and unit tests now.

**Comment by pcd1193182** [comment]
I went through and copied over all  the diffs I thought were correct; all but the `/* the following fields are reserved for redacted send / recv */` line, which is now out of date since redacted send/recv has landed.

**Pull request merged**